### ISSUE TITLE
[xcat-inventory osimage] As a user, how to use xcat-inventory osimage  management

### ISSUE THREAD

**Issue by zet809** [xcat2/xcat2-task-management#50]
As a user, how to use xcat-inventory to do osimage definition+files(pkglist, script, template)+binary(rpm/deb package) management

**Comment by whowutwut** [comment]
Is there a use case about the site table config and how to roll back before we start this one?

**Comment by gurevichmark** [comment]
For reference: https://github.ibm.com/vhu/c910env/issues/247

**Comment by whowutwut** [comment]
@zet809 This use case is way too large...  It should be a smaller use case: 

* As a admin of xCAT, I want to roll back the configuration data to my configuration 2 weeks ago
* As an admin of xCAT, I want to roll back to a previous configuration of my osimage

I think the first thing that needs to change is the ability to source control site config separate from osimage config, then we can focus on a smaller user case,  osimage may change very frequently, site config may not....

**Comment by immarvin** [comment]
The link to the draft work flow of osimage version control

https://ibm.box.com/s/ooqz3nifd6wq686kwg3t3vn97324csqf

**Comment by immarvin** [comment]
User Case of osimage version control:
```
[root@c910f03c05k21 osimage]# lsdef -t osimage -o rhels7.4-x86_64-netboot-compute
Object name: rhels7.4-x86_64-netboot-compute
    exlist=/opt/inventory/site/osimage/rhels7.4-x86_64-netboot-compute/compute.rhels7.x86_64.exlist
    imagetype=linux
    osarch=x86_64
    osdistroname=rhels7.4-x86_64
    osvers=rhels7.4
    otherpkgdir=/install/post/otherpkgs/rhels7.4/x86_64
    otherpkglist=/opt/inventory/site/osimage/rhels7.4-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist
    pkgdir=/install/rhels7.4/x86_64
    pkglist=/opt/inventory/site/osimage/rhels7.4-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
    postinstall=/opt/inventory/site/osimage/rhels7.4-x86_64-netboot-compute/compute.rhels7.x86_64.postinstall
    profile=compute
    provmethod=netboot
    rootimgdir=/install/netboot/rhels7.4/x86_64/compute

[root@c910f03c05k21 osimage]# xcat-inventory export -t osimage -o rhels7.4-x86_64-netboot-compute --path /opt/inventory/site/osimage/

[root@c910f03c05k21 osimage]# tree /opt/inventory/site/osimage/
/opt/inventory/site/osimage/
└── rhels7.4-x86_64-netboot-compute
    ├── compute.rhels7.x86_64.exlist
    ├── compute.rhels7.x86_64.otherpkglist
    ├── compute.rhels7.x86_64.pkglist
    ├── compute.rhels7.x86_64.postinstall
    └── def.yaml

1 directory, 5 files

[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# cat ./def.yaml
osimage:
  rhels7.4sp1-x86_64-netboot-compute:
    basic_attributes:
      arch: x86_64
      distribution: rhels7.4sp1
      osdistro: rhels7.4sp1-x86_64
    genimgoptions:
      exlist: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.exlist
      postinstall: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.postinstall
      rootimgdir: /install/netboot/rhels7.4sp1/x86_64/compute
    imagetype: linux
    package_selection:
      otherpkgdir: /install/post/otherpkgs/rhels7.4sp1/x86_64
      otherpkglist: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist
      pkgdir: /install/rhels7.4sp1/x86_64
      pkglist: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
    provision_mode: netboot
    role: compute
schema_version: '1.0'

[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# xcat-inventory import -t osimage -f ./def.yaml
Importing object: rhels7.4sp1-x86_64-netboot-compute
Inventory import successfully!
[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# lsdef -t osimage -o rhels7.4sp1-x86_64-netboot-compute
Object name: rhels7.4sp1-x86_64-netboot-compute
    exlist=/opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.exlist
    imagetype=linux
    osarch=x86_64
    osdistroname=rhels7.4sp1-x86_64
    osvers=rhels7.4sp1
    otherpkgdir=/install/post/otherpkgs/rhels7.4sp1/x86_64
    otherpkglist=/opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist
    pkgdir=/install/rhels7.4sp1/x86_64
    pkglist=/opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
    postinstall=/opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.postinstall
    profile=compute
    provmethod=netboot
    rootimgdir=/install/netboot/rhels7.4sp1/x86_64/compute

[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]#  git add ./rhels7.4sp1-x86_64-netboot-compute/
[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]#  git commit ./rhels7.4sp1-x86_64-netboot-compute/ -m "add new osimage definition rhels7.4sp1-x86_64-netboot-compute for rhels7.4-x86_64  SP1"
[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]#  git push origin
```


```
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# git pull
Receiving objects: 100% (24/24), done.
Resolving deltas: 100% (4/4), done.
```

modify osimage definition `rhels7.4sp1-x86_64-netboot-compute`,
```
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# git diff
diff --git a/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist b/site/osimage/rhels7.4sp1-x86_64-
index e69de29..a0ef0e4 100644
--- a/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist
+++ b/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist
@@ -0,0 +1 @@
+cuda-repo-rhel7
diff --git a/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist b/site/osimage/rhels7.4sp1-x86_64-netbo
index e69de29..a842855 100644
--- a/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
+++ b/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
@@ -0,0 +1,3 @@
+vim
+bc
+ruby

[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# xcat-inventory import -f ./def.yaml
Importing object: rhels7.4sp1-x86_64-netboot-compute
Inventory import successfully!
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# genimage rhels7.4sp1-x86_64-netboot-compute
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# rinstall node01 osimage=rhels7.4sp1-x86_64-netboot-compute
```

the osimage verification passed, commit the osimage 
```
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# git add compute.rhels7.x86_64.otherpkglist compute.rhels7.x86_64.otherpkglist
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# git commit -m "add several toolkit into pkglist; add cuda tookit into otherpkglist"
[root@briggs01 rhels7.4sp1-x86_64-netboot-compute]# git push origin
```

**Comment by immarvin** [comment]
ping @whowutwut , please take a look at the use cases , any comment is appreciated, thanks

**Comment by whowutwut** [comment]
@immarvin This is a little better, we are moving in the right direction! 👍  

I still think as we do this exercise, you should provision a real node, then you will see some of the gaps. 
For example, when you import the osimage into briggs01...  the files are not there, how would those files be kept in sync? They are exported to `/opt/inventory/site/osimage/` but that is not part of the source control, so there's a gap here. 

```
[root@briggs01 inventory-test]# ls -ltr /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
ls: cannot access /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist: No such file or directory
[root@briggs01 inventory-test]#
```

So this would not successfully provision.  

You are also still using `root` user on the machines, while this is easier, it's a big security hole,  Look at what I could do... https://github.com/immarvin/inventory-test/commit/c02d8ebdb6a811382e95bd835237ba5457466751
```
[root@briggs01 inventory-test]# vi ./site/osimage/rhels7.4-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
[root@briggs01 inventory-test]# git add ./site/osimage/rhels7.4-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
[root@briggs01 inventory-test]# git commit -m "Hacked..."
[master c02d8eb] Hacked...
 1 file changed, 4 insertions(+)
[root@briggs01 inventory-test]# git push origin master
Counting objects: 11, done.
Delta compression using up to 128 threads.
Compressing objects: 100% (4/4), done.
Writing objects: 100% (6/6), 558 bytes | 0 bytes/s, done.
Total 6 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To git@github.com:immarvin/inventory-test.git
   a952660..c02d8eb  master -> master
[root@briggs01 inventory-test]# pwd
/home/yangsbj/git/inventory-test
```

So essentially on a shared machine, you allowed me access to your account, this is a big no no.  


I also think when we start looking at creating new osimages vs maintaining osimages, the use case completely changes.  In maintenance, there's really nothing changing in the osimage, but rather in the postscripts, pkglist, postinstall .... so those files would be tagged and we could checkout the tag to revert the image...   

by checking out the tag, the files (pkglist, postinstall postscripts, ) will change to the older version, simply by launching the osimage or running `genimage` will create a new image, There's no need to import again.

**Comment by immarvin** [comment]
@whowutwut , I think your comments can be summarized to 3 questions:
1. how to import **customized osimage files**(“pkglist, template, otherpkglist, partition file/script, exlist, postinstall”) from the osimage source control directory to the path in osimage definition?

2. how to control and audit the access of git repo on a shared environment with multiple Admins?

3. if the osimage definition does not change, is it possible to switch the osimage versions by simply `git checkout <commit>`?

**Comment by immarvin** [comment]
>  how to import customized osimage files(“pkglist, template, otherpkglist, partition file/script, exlist, postinstall”) from the osimage source control directory to the path in osimage definition?

We have 2 solutions here:

**solution 1**:  preserve the full paths of the customized osimage files under exported osimage directory under source control during `xcat-inventory export`, and restore them to the defined full path during `xcat-inventory import`, an example: 

for the osimage "rhels7.4sp1-x86_64-netboot-compute": 
```
[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# cat ./def.yaml
osimage:
  rhels7.4sp1-x86_64-netboot-compute:
    basic_attributes:
      arch: x86_64
      distribution: rhels7.4sp1
      osdistro: rhels7.4sp1-x86_64
    genimgoptions:
      exlist: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.exlist
      postinstall: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.postinstall
      rootimgdir: /install/netboot/rhels7.4sp1/x86_64/compute
    imagetype: linux
    package_selection:
      otherpkgdir: /install/post/otherpkgs/rhels7.4sp1/x86_64
      otherpkglist: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist
      pkgdir: /install/rhels7.4sp1/x86_64
      pkglist: /opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.pkglist
    provision_mode: netboot
    role: compute
    scripts:
      postbootscripts: config_cuda
      postscripts: confignetwork
schema_version: '1.0'
```

the exported directory under source control will look like:
```
[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# xcat-inventory export -t osimage -o rhels7.4sp1-x86_64-netboot-compute --format=yaml --path /opt/inventory/site/osimage/

 [root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# tree
.
├── config_cuda
├── confignetwork
├── def.yaml
└── opt
    └── inventory
        └── site
            └── osimage
                └── rhels7.4sp1-x86_64-netboot-compute
                    ├── compute.rhels7.x86_64.exlist
                    ├── compute.rhels7.x86_64.otherpkglist
                    ├── compute.rhels7.x86_64.pkglist
                    └── compute.rhels7.x86_64.postinstall

```

the file `opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist` under source control will be copied to `/opt/inventory/site/osimage/rhels7.4sp1-x86_64-netboot-compute/compute.rhels7.x86_64.otherpkglist` after `xcat-inventory import`, will overwrite if already exists. User need to make sure there is no confliction when multiple osimages share the same customized files.

**solution 2**:  support osimage working directory and relative path in osimage definition. With the 

The new prototype of osimage looks like : 
```
[root@c910f03c05k21 rhels7.4sp1-x86_64-netboot-compute]# cat ./def.yaml
osimage:
  rhels7.4sp1-x86_64-netboot-compute:
    basic_attributes:
      arch: x86_64
      distribution: rhels7.4sp1
      osdistro: rhels7.4sp1-x86_64
    genimgoptions:
      exlist: ./compute.rhels7.x86_64.exlist
      postinstall: ./compute.rhels7.x86_64.postinstall
      rootimgdir: /install/netboot/rhels7.4sp1/x86_64/compute
    imagetype: linux
    package_selection:
      otherpkgdir: ./x86_64
      otherpkglist: ./compute.rhels7.x86_64.otherpkglist
      pkgdir: /install/rhels7.4sp1/x86_64
      pkglist: ./compute.rhels7.x86_64.pkglist
    provision_mode: netboot
    role: compute
schema_version: '1.0'

[root@c910f03c05k21 osimage]# tree /opt/inventory/site/osimage/
/opt/inventory/site/osimage/
└── rhels7.4sp1-x86_64-netboot-compute
    ├── compute.rhels7.x86_64.exlist
    ├── compute.rhels7.x86_64.otherpkglist
    ├── compute.rhels7.x86_64.pkglist
    ├── compute.rhels7.x86_64.postinstall
    └── def.yaml
```
each osimage has a working directory under `/install/osimages/<osimage name>`,  this directory contains all the files with relative paths in osimage definition, xcat will refer the files with relative paths based on the working directory. 

After the osimage is imported by `xcat-inventory import`, all the customized osimages files under the git directory will be copied to the working directory.  with this support, user do not need to care about the conflict for the osimages with relative paths

**Comment by immarvin** [comment]
> 2. how to control and audit the access of git repo on a shared environment with multiple Admins?

When multiple admins shares one management node, we suggest each admin have his/her own local fork of the source control repo, admin A has `/home/A/inventory/`, B has `/home/B/inventory`, the access control of the OS guarantees each admin have correct permission over his/her local fork, and each admin's commit has correct correct audit info.

**Comment by immarvin** [comment]
> 3. if the osimage definition does not change, is it possible to switch the osimage versions by simply git checkout <commit>?

this is feasible if the path the osimage customized files refer to the files under source control, but`xcat-inventory import` is mandatory if the osimage definition changed,  and it introduces additional effort for the user to tell  whether the osimage definition, which is error prone.

**Comment by immarvin** [comment]
move the discussion on the postscript source control here: 
```
@yangsbj But this is a complete paradigm shift from what xCAT does today.  What is the technical reason for that?

`/install/postscripts` is the top level and allow the admins to decide themselves where to put the files…dictate where to put the files.. for example, they might have a library of scripts for mellanox…   or even put their files somewhere else and symlink it to the `/install/postscripts`

`/install/postscripts/mellanox/<scripts>`  and they can reference it like `updatenode -P <noderange> mellanox/scriptabc` ..  by doing the above, we are changing the understood behavior of postscripts… (edited)
```

the situation becomes little complicated after we introduce osimage source control and sharing, one scenario,  in the working environment, user created customized postscripts "A","B","C" under `/install/postscripts`,  these customized postscripts should be source controlled, `xcat-inventory export ` should grab these files to the source control directory, but how can we tell whether these postscripts are customized or shipped by xCAT? maintain a list of xCAT shipped postscripts?  we might need to source control all the postscripts referred in osimage definition. 

During osimage import, If we simply source controlled it and copy them to `/install/postscripts/`, there might be conflict between the existing customized postscript with the same name.

The idea I mentioned in slack is just some thought of us, might need additional effort and can be low priority.

another more feasible idea is, while importing the osimage, first `diff` the files under source control and any existing files with same name under `/install/postscripts`,  skip copy if the 2 are exactly the same; otherwise,  throw out a message to user and let the user to determine whether to overwrite or solve the conflict manually.

**Comment by immarvin** [comment]
latest proposal https://github.ibm.com/hpc-devops/xcat_clusters/pull/33

**Comment by immarvin** [comment]
the demo finished, close this ticket
